**Por que estatística é importante em machine learning?**

Um modelo aprende padrões a partir de dados, isso difere da programação comum em que o humano insere REGRAS DETERMINADAS padrões para um programa realizar uma tarefa.

A estatística é ferramenta para:
- Para entender os dados antes
- Percebermos e resolvermos problemas como valores estranhos ou extremos, dados faltantes ou mal registrados
- Comparar previsão com o valor real
- Analisar o quanto determinado resultado é confiável
- Evitar conclusões preciptadas sobre o que modelo aprendeu

**Analise exploratória e validação estatisitica**

Por que?
Conhecer os dados antes e fazer uso da estatística nos permite uma orientação melhor sobre o modelo
Modelos complexos para problemas simples ou modelos simples para problemas complexos geram perda de tempo, recurso e consequentemente valor no desenvolvimento de um projeto de IA

**Aplicação**
Vamos explorar a base de dados, *base_imoveis.csv*
Primeiro começamos lendo a base para criar o dataframe, realizar contagens, explorar a base.
Depois vamos restringir o dataframe para apenas algumas linhas e realizar a validação estatística afim de nos orientarmos sobre o modelo de ML
Plot de gráficos pra auxiliar a visualização
Por fim teremos a justificativa de uso do modelo

**LEITURA E RECONHECIMENTO**

In [4]:
import pandas as pd

# 1. Carregar a base de dados
df = pd.read_csv(
    r'C:\Users\cristiano59293196\Documents\GITHUB\ml_stat\bases_dados\base_imoveis.csv',
    encoding='utf-8'
)

# 2. Visualizar as primeiras linhas
display(df.head())

# 3. Verificar o tamanho da base (linhas, colunas)
print(df.shape)

# 4. Mapear os tipos de colunas e dados ausentes
print(df.info())

# 5. Estatística descritiva inicial (para variáveis numéricas)
print("\n--- Resumo Estatístico Inicial ---")
display(df.describe())


,id_imovel,localidade,tipo,area_m2,quartos,banheiros,vagas_garagem,idade_anos,preco_venda
0,IMV-10000,Candangolândia,Apartamento,152.0,2.0,2.0,2.0,24.0,468757.0
1,IMV-10001,Riacho Fundo,Casa,126.0,4.0,5.0,2.0,30.0,366019.0
2,IMV-10002,Riacho Fundo,Casa,587.0,4.0,4.0,4.0,36.0,1777368.0
3,IMV-10003,Gama,Apartamento,78.0,3.0,2.0,1.0,14.0,262673.0
4,IMV-10004,Cabeceira Grande,Casa,255.0,NaN,4.0,2.0,3.0,374877.0


(5500, 9)
<class 'pandas.DataFrame'>
RangeIndex: 5500 entries, 0 to 5499
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_imovel      5500 non-null   str    
 1   localidade     5210 non-null   str    
 2   tipo           5200 non-null   str    
 3   area_m2        5216 non-null   float64
 4   quartos        5156 non-null   float64
 5   banheiros      5182 non-null   float64
 6   vagas_garagem  5181 non-null   float64
 7   idade_anos     5180 non-null   float64
 8   preco_venda    5173 non-null   float64
dtypes: float64(6), str(3)
memory usage: 386.8 KB
None

--- Resumo Estatístico Inicial ---


,area_m2,quartos,banheiros,vagas_garagem,idade_anos,preco_venda
count,5216.000000,5156.000000,5182.000000,5181.000000,5180.000000,5.173000e+03
mean,294.827837,3.621412,4.149170,2.566879,20.087645,8.732209e+05
std,163.986606,1.113622,1.792976,1.071757,11.778055,7.552051e+05
min,40.000000,1.000000,1.000000,0.000000,0.000000,7.895600e+04
25%,148.000000,3.000000,3.000000,2.000000,10.000000,4.036640e+05
50%,275.000000,4.000000,4.000000,3.000000,20.000000,6.653100e+05
75%,439.000000,5.000000,5.000000,3.000000,30.000000,1.076094e+06
max,600.000000,5.000000,7.000000,4.000000,40.000000,7.373758e+06


**Perguntas Inciais**

- Quais são dimensões da base? 
Shape: (5500, 9)

- Quantas features existem? Existem 9 features
id_imovel
localidade
tipo
area_m2
quartos
banheiros
vagas_garagem
idade_anos
preco_venda

- Quais são categoricas e quais contínuas?

Variáveis categóricas

localidade
tipo

Variáveis numéricas

area_m2
quartos
banheiros
vagas_garagem
idade_anos
preco_venda

**Missing Values**
- Existem dados ausentes? Se sim, em quais colunas?
localidade
tipo
area_m2
quartos
banheiros
vagas_garagem
idade_anos
preco_venda

- Como esses dados ausentes afetam o treinamento de um modelo de ML?
Impedir o treinamento de alguns algoritmos.
Reduzir a qualidade das previsões.
Introduzir vieses nos resultados.
Diminuir a representatividade da amostra.

Gerar conclusões incorretas sobre os padrões dos dados.
-  Para discussão: O que podemos fazer com esses dados ausentes:
    -   Inserindo artificalmente a média ou mediana quais consequencias para o modelo?
    - Exclusões quais seriam as consequências para o modelo?

**Resumo Estatístico**
- Qual maior e menor preço?
Maior preço: R$ 7.373.758
Menor preço: R$ 78.956

- Qual é média de idade dos imóveis?
A média de idade dos imóveis é de aproximadamente:
20,09 anos

- Qual faixa percentual encontra-se a maior metragem dos imóveis

Percentis da variável area_m2:

Percentil	Área (m²)
25%	           148
50%	           275
75%	           439
- Qual é a mediana do preço de venda?

R$ 665.310

- Qual percentual custa menos de 665k?

46,98%

**Dispersão**
Percentuais altos indicam alta **HETEROGENIDADE**
- Vamos calcular a relação entre o desvio padrao e a média do preço e tirar conclusões

In [5]:
df['preco_venda'].std()
df['preco_venda'].mean()
dispersao = df['preco_venda'].std() / df['preco_venda'].mean()*100
print(f"\n--- Dispersão (Desvio Padrão / Média) ---\nDispersão: {dispersao:.2f}%")


--- Dispersão (Desvio Padrão / Média) ---
Dispersão: 86.49%


**Conclusão mateática sobre a base**
**É altamente heteorgenea**
- Vamos restringir o df e focar em uma ou 2 features
- vamos contabilizar os nulos e entender o percentual de nulos da base
- apartamentos e casas tem tamanhos geralmente diferentes o que pode influenciar ou não no preço...
- assim vamos partir os dados entre casas e aprtamentos, eliminar os valores nulos e analisar os resultados comparando segmentados com nao segmentados

In [6]:
total_nulos = df.isnull().sum().sum()
percentual_nulos = (total_nulos / df.size) * 100

print(f"\n--- Percentual de Valores Nulos na Base ---\nTotal de valores nulos: {total_nulos}\nPercentual de valores nulos: {percentual_nulos:.2f}%")

print(f"\n--- Percentual de nulos por coluna---\n{((df.isnull().sum() / len(df)) * 100)}")



--- Percentual de Valores Nulos na Base ---
Total de valores nulos: 2502
Percentual de valores nulos: 5.05%

--- Percentual de nulos por coluna---
id_imovel        0.000000
localidade       5.272727
tipo             5.454545
area_m2          5.163636
quartos          6.254545
banheiros        5.781818
vagas_garagem    5.800000
idade_anos       5.818182
preco_venda      5.945455
dtype: float64


In [9]:
#dataframses segmentados (removendo nulos do preço)
df_apartamentos = df[df['tipo'] == 'Apartamento'].dropna(subset=['preco_venda'])
df_casas = df[df['tipo'] == 'Casa'].dropna(subset=['preco_venda'])

In [8]:
print("=== APARTAMENTOS ===")
print (df_apartamentos.describe())

=== APARTAMENTOS ===
           area_m2      quartos    banheiros  vagas_garagem   idade_anos  \
count  1186.000000  1164.000000  1164.000000    1171.000000  1171.000000   
mean    108.826307     2.505155     1.746564       1.293766    20.117848   
std      39.771120     1.103021     0.913375       0.645129    11.698920   
min      40.000000     1.000000     1.000000       0.000000     0.000000   
25%      75.000000     2.000000     1.000000       1.000000    10.000000   
50%     109.000000     3.000000     1.000000       1.000000    20.000000   
75%     142.000000     3.000000     2.000000       2.000000    31.000000   
max     180.000000     4.000000     4.000000       2.000000    40.000000   

        preco_venda  
count  1.243000e+03  
mean   5.371363e+05  
std    3.898926e+05  
min    7.895600e+04  
25%    2.573660e+05  
50%    4.033000e+05  
75%    7.136465e+05  
max    2.101933e+06  


In [10]:
print("=== CASAS ===")
print (df_casas.describe())

=== CASAS ===
           area_m2      quartos    banheiros  vagas_garagem   idade_anos  \
count  3461.000000  3431.000000  3447.000000    3454.000000  3432.000000   
mean    360.253395     4.005246     4.975341       3.010423    19.937646   
std     139.067619     0.817431     1.161496       0.808230    11.795754   
min     120.000000     3.000000     3.000000       2.000000     0.000000   
25%     238.000000     3.000000     4.000000       2.000000    10.000000   
50%     360.000000     4.000000     5.000000       3.000000    20.000000   
75%     481.000000     5.000000     6.000000       4.000000    30.000000   
max     600.000000     5.000000     7.000000       4.000000    40.000000   

        preco_venda  
count  3.653000e+03  
mean   9.923047e+05  
std    8.213981e+05  
min    1.234470e+05  
25%    4.939330e+05  
50%    7.607460e+05  
75%    1.195266e+06  
max    7.373758e+06  


In [12]:
#teste de normalidade (Shapiro-Wilk) para verificar se os preços de venda seguem uma distribuição normal
from scipy import stats
_, p_ap = stats.shapiro(df_apartamentos['preco_venda'].sample(min(len(df_apartamentos), 1000), random_state=42))
_, p_ca = stats.shapiro(df_casas['preco_venda'].sample(min(len(df_casas), 1000), random_state=42))


In [13]:
print("\n--- Teste de Normalidade (Amostra de 1000 imóveis) ---")
print(f"Apartamentos - p-valor: {p_ap:.4f} -> " + ("É Normal" if p_ap > 0.05 else "NÃO é Normal"))
print(f"Casas        - p-valor: {p_ca:.4f} -> " + ("É Normal" if p_ca > 0.05 else "NÃO é Normal"))


--- Teste de Normalidade (Amostra de 1000 imóveis) ---
Apartamentos - p-valor: 0.0000 -> NÃO é Normal
Casas        - p-valor: 0.0000 -> NÃO é Normal


**Teste Shapiro Wilk**

O teste é indicador de normalidade, ou seja que a distribuição dos dados é equalitaria no histograma;
Ou em outras palavras se o resultado é normal, então **podemos confiar na média** não há muitas distorções e o modelo de machine learning\
de regressão linear simples pode ser usado.
**p valor não normal** Aí siginifca que há outiliers sigfificativos e que o modelo de arvores de decisão é o mais adequado, como random forest